### PR TITLE
Docker toolchain autoconfiguration using 'which'

### PR DESCRIPTION
**Pull request by smukherj1**
Also updated docs

### CODE REVIEW

File: toolchains/docker/readme.md [hunk 1]
```diff
@@ -28,6 +28,10 @@ register_toolchains(
 )
 ```
 
+The docker toolchain automatically configures itself to use the docker in your
+system path. To override this behavior, you need to call explicitly call the
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply toolchains/docker/readme.md]: Done

File: toolchains/docker/toolchain.bzl [hunk 2]
```diff
@@ -30,9 +30,28 @@ def _docker_toolchain_impl(ctx):
     )
     return [toolchain_info]
 
+# Regular rule used by the docker toolchain rule to specify a path to a
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

File: toolchains/docker/toolchain.bzl [hunk 3]
```diff
@@ -30,9 +30,28 @@ def _docker_toolchain_impl(ctx):
...
+# particular docker toolchain
 docker_toolchain = rule(
     implementation = _docker_toolchain_impl,
     attrs = {
         "tool_path": attr.string(),
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

File: toolchains/docker/toolchain.bzl [hunk 4]
```diff
@@ -30,9 +30,28 @@ def _docker_toolchain_impl(ctx):
...
+    repository_ctx.template(
+        "BUILD",
+        Label("@io_bazel_rules_docker//toolchains/docker:BUILD.tpl"),
+        {
+            "%{DOCKER_TOOL}": "%s" % tool_path,
+        },
+        False,
+    )
+
+# Repository rule to automatically generate a docker toolchain rule using
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

File: container/container.bzl [hunk 5]
```diff
@@ -219,3 +223,8 @@ py_library(
         "@io_bazel_rules_docker//toolchains/docker:default_windows_toolchain",
         "@io_bazel_rules_docker//toolchains/docker:default_osx_toolchain",
     )
+
+    if "local_docker_config" not in excludes:
+        # Automatically configure the docker toolchain rule with the default
+        # docker toolchain from the system path
+        _docker_toolchain_configure(name = "local_docker_config")
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply container/container.bzl]: Done

File: toolchains/docker/BUILD [hunk 6]
```diff
@@ -34,7 +34,7 @@ toolchain(
         "@bazel_tools//platforms:linux",
         "@bazel_tools//platforms:x86_64",
     ],
-    toolchain = ":default_toolchain_impl",
+    toolchain = "@local_docker_config//:toolchain",
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply toolchains/docker/BUILD]: Done

File: container/container.bzl [hunk 7]
```diff
@@ -219,3 +223,8 @@ py_library(
         "@io_bazel_rules_docker//toolchains/docker:default_windows_toolchain",
         "@io_bazel_rules_docker//toolchains/docker:default_osx_toolchain",
     )
+
+    if "local_docker_config" not in excludes:
+        # Automatically configure the docker toolchain rule with the default
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply container/container.bzl]: Done

File: README.md [hunk 8]
```diff
@@ -91,6 +91,13 @@ http_archive(
     urls = ["https://github.com/bazelbuild/rules_docker/archive/v0.5.1.tar.gz"],
 )
 
+# Call this to override the default docker toolchain configuration. This
+# be called BEFORE calling "container_repositories" below to actually override
+# the default toolchain configuration
+load("@io_bazel_rules_docker//toolchains/docker:toolchain.bzl",
+    docker_toolchain_configure="toolchain_configure"
+)
+docker_toolchain_configure(name = "local_docker_config")
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply README.md]: Done

File: README.md [hunk 9]
```diff
@@ -91,6 +91,13 @@ http_archive(
     urls = ["https://github.com/bazelbuild/rules_docker/archive/v0.5.1.tar.gz"],
 )
 
+# Call this to override the default docker toolchain configuration. This
+# be called BEFORE calling "container_repositories" below to actually override
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply README.md]: Done

File: toolchains/docker/toolchain.bzl [hunk 10]
```diff
@@ -30,9 +30,29 @@ def _docker_toolchain_impl(ctx):
     )
     return [toolchain_info]
 
+# Regular rule used by the docker toolchain rule to specify a path to the docker
```
Comment (nlopezgi) [inline toolchains/docker/toolchain.bzl]: r/ Regular

File: toolchains/docker/toolchain.bzl [hunk 11]
```diff
@@ -30,9 +30,29 @@ def _docker_toolchain_impl(ctx):
...
+    repository_ctx.template(
+        "BUILD",
+        Label("@io_bazel_rules_docker//toolchains/docker:BUILD.tpl"),
+        {
+            "%{DOCKER_TOOL}": "%s" % tool_path,
+        },
+        False,
+    )
+
+# Repository rule to automatically generate a docker_toolchain target
```
Comment (nlopezgi) [inline toolchains/docker/toolchain.bzl]: r/ automatically

File: toolchains/docker/readme.md [hunk 12]
```diff
@@ -28,6 +28,10 @@ register_toolchains(
 )
 ```
 
+The docker toolchain automatically configures itself to use the docker in your
```
Comment (nlopezgi) [inline toolchains/docker/readme.md]: These rules by default use the docker in your system path. To override...

File: toolchains/docker/toolchain.bzl (hunk 10 above)
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

File: toolchains/docker/readme.md (hunk 12 above)
Comment (smukherj1) [reply toolchains/docker/readme.md]: Done

File: toolchains/docker/toolchain.bzl (hunk 11 above)
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

**Comment by nlopezgi** [comment]
ugh, all of our mac CI builds are timing out downloading images, wonder what's up

**Comment by smukherj1** [comment]
> ugh, all of our mac CI builds are timing out downloading images, wonder what's up

Seems to be resolved now. I retried the build and it was fine

**Pull request merged**